mobile_verifier/src/reward_shares.rs

```diff
@@ -3,17 +3,132 @@ use crate::{
+const DEFAULT_PREC: u32 = 15;
+
+pub struct TransferRewards {
+    scale: Decimal,
+    rewards: HashMap<PublicKeyBinary, Decimal>,
+    pub remaining_poc_rewards: Decimal,
+}
+
+impl TransferRewards {
+    pub fn empty(epoch: &Range<DateTime<Utc>>) -> Self {
+        Self {
+            scale: Decimal::ONE,
+            rewards: HashMap::new(),
+            remaining_poc_rewards: get_scheduled_tokens(epoch.start, epoch.end - epoch.start)
+                .unwrap(),
+        }
+    }
+
+    pub fn scale(&self) -> Decimal {
+        self.scale
+    }
+
+    pub fn reward(&self, hotspot: &PublicKeyBinary) -> Decimal {
+        self.rewards.get(hotspot).copied().unwrap_or(Decimal::ZERO) * self.scale
+    }
+
+    pub async fn from_transfer_sessions(
+        mobile_price: Decimal,
+        transfer_sessions: impl Stream<Item = ValidDataTransferSession>,
+        epoch: &Range<DateTime<Utc>>,
+    ) -> Result<Self, ClockError> {
+        tokio::pin!(transfer_sessions);
+
+        let mut data_transfer_reward_sum = Decimal::ZERO;
+        let rewards = transfer_sessions
+            // Accumulate bytes per hotspot
+            .fold(
+                HashMap::<PublicKeyBinary, Decimal>::new(),
+                |mut entries, session| async move {
+                    *entries.entry(session.pub_key).or_default() +=
+                        Decimal::from(bytes_to_dc(session.download_bytes + session.upload_bytes));
+                    entries
+                },
+            )
+            .await
+            .into_iter()
+            // Calculate rewards per hotspot
+            .map(|(pub_key, dc_amount)| {
+                let mobiles = dc_to_mobile_bones(dc_amount, mobile_price);
+                data_transfer_reward_sum += mobiles;
+                (pub_key, mobiles)
+            })
+            .collect();
+
+        // Determine if we need to scale the rewards given for data transfer rewards.
+        // Ideally this should never happen, but if the total number of data transfer rewards
+        // is greater than (at the time of writing) 40% of the total pool, we need to scale
+        // the rewards given for data transfer.
+        //
+        // If we find that total data_transfer reward sum is greater than 80%, we use the
+        // following math to calculate the scale:
+        //
+        // [ scale * data_transfer_reward_sum ] / total_emissions_pool = 0.4
+        //
+        //   therefore:
+        //
+        // scale = [ 0.4 * total_emissions_pool ] / data_transfer_reward_sum
+        //
+        let scale = if data_transfer_reward_sum / TOTAL_EMISSIONS_POOL
+            > MAX_DATA_TRANSFER_REWARDS_PERCENT
+        {
+            MAX_DATA_TRANSFER_REWARDS_PERCENT * TOTAL_EMISSIONS_POOL / data_transfer_reward_sum
+        } else {
+            Decimal::ONE
+        };
+        let remaining_poc_rewards = get_scheduled_tokens(epoch.start, epoch.end - epoch.start)?
+            - scale * data_transfer_reward_sum;
```

Comment:
```suggestion
        let remaining_poc_rewards = get_scheduled_tokens(epoch.start, epoch.end - epoch.start)?
            - (scale * data_transfer_reward_sum);
```

Reply:
should this guard against `remaining_poc_rewards` being a negative number in case we're missing some edge case in the math?